feat(map.jinja): load configuration values from configurable sources (#186)
saltstack-formulas/template-formula: docs/map.jinja.rst

.. _map.jinja:

``map.jinja``: gather formula configuration values
==================================================

The `documentation`_ gives a quick view of the use of a ``map.jinja`` to gather parameters values for a formula.

> The today best practice is to let ``map.jinja`` handle parameters from all sources and avoid the call of pillars, grains or configuration from ``sls`` files and template directly.

Reviewer: ```suggestion
Current best practice is to let ``map.jinja`` handle parameters from all sources, to minimise the use of pillars, grains or configuration from ``sls`` files and templates directly.
```

.. contents:: **Table of Contents**

How to set configuration values of a formula
--------------------------------------------

The ``map.jinja`` use several sources where to lookup parameter values:

- YAML files located in the `fileserver`_ under the formula ``parameters/`` directory
- configuration collected by `salt['config.get']`_

As `pillars`_ are rendered by the SaltStack master for each minion, they could became costly when you have lots of minions with many pillar values.

As a good practice, you should:

- store non secret data in YAML files distributed by the `fileserver`_
- store sensible data

  - in `pillars`_ (and look for the use of something like `pillar.vault`_)

>   - in `SDB`_ (and look for the use of something like `sbd.vault`_)

Reviewer: ```suggestion
The ``map.jinja`` file uses several sources to lookup parameter values:

- Multiple YAML files organised in the `fileserver`_ under the formula ``parameters/`` directory
- Configuration collected by `salt['config.get']`_

As `pillars`_ are rendered on the Salt master for every minion, this increases the load on the master as the pillar values and the number of minions grow.

As a good practice, you should:

- store non-secret data in YAML files distributed by the `fileserver`_
- store secret data in:

  - `pillars`_ (and look for the use of something like `pillar.vault`_)
  - `SDB`_ (and look for the use of something like `sdb.vault`_)
```

Author: Great, much better. And I always do the `sdb` / `sbd` typo :-/

            TEMPLATE: {{ TEMPLATE | json }}

This ``sls`` file expose a ``TEMPLATE`` context variable to the jinja template which could be used like this:

.. code-block:: jinja

    ########################################################################
    # File managed by Salt at <{{ source }}>.
    # Your changes will be overwritten.
    ########################################################################

    This is another example file from SaltStack template-formula.

    # This is here for testing purposes
    {{ TEMPLATE | json }}

    winner of the merge: {{ TEMPLATE['winner'] }}

.. _documentation: https://docs.saltstack.com/en/latest/topics/development/conventions/formulas.html#writing-formulas
.. _fileserver: https://docs.saltstack.com/en/latest/ref/configuration/master.html#std:conf_master-fileserver_backend
.. _salt['config.get']: https://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.config.html#salt.modules.config.get
.. _pillar.vault: https://docs.saltstack.com/en/latest/ref/pillar/all/salt.pillar.vault.html
.. _pillars: https://docs.saltstack.com/en/latest/topics/pillar/
.. _SDB: https://docs.saltstack.com/en/latest/topics/sdb/index.html

> .. _sbd.vault: https://docs.saltstack.com/en/latest/ref/sdb/all/salt.sdb.vault.html

Reviewer: As fixed above:

```suggestion
.. _sdb.vault: https://docs.saltstack.com/en/latest/ref/sdb/all/salt.sdb.vault.html
```

Author: again…
